plugins/node/instrumentation-kafkajs/src/instrumentation.ts

```diff
@@ -0,0 +1,452 @@
+          }
+        );
+      };
+    };
+  }
+
+  private _getProducerSendBatchPatch() {
+    const instrumentation = this;
+    return (original: Producer['sendBatch']) => {
+      return function sendBatch(
+        this: Producer,
+        ...args: Parameters<Producer['sendBatch']>
+      ): ReturnType<Producer['sendBatch']> {
+        const batch = args[0];
+        const messages = batch.topicMessages || [];
+        const spans: Span[] = messages
+          .map(topicMessage =>
+            topicMessage.messages.map(message =>
+              instrumentation._startProducerSpan(topicMessage.topic, message)
+            )
+          )
+          .reduce((acc, val) => acc.concat(val), []);
+
+        const origSendResult: Promise<RecordMetadata[]> = original.apply(
+          this,
+          args
+        );
+        return instrumentation._endSpansOnPromise(spans, origSendResult);
+      };
+    };
+  }
+
+  private _getProducerSendPatch() {
+    const instrumentation = this;
+    return (original: Producer['send']) => {
+      return function send(
+        this: Producer,
+        ...args: Parameters<Producer['send']>
+      ): ReturnType<Producer['send']> {
+        const record = args[0];
+        const spans: Span[] = record.messages.map(message => {
+          return instrumentation._startProducerSpan(record.topic, message);
+        });
+
+        const origSendResult: Promise<RecordMetadata[]> = original.apply(
+          this,
+          args
+        );
+        return instrumentation._endSpansOnPromise(spans, origSendResult);
+      };
+    };
+  }
+
+  private _endSpansOnPromise<T>(
+    spans: Span[],
+    sendPromise: Promise<T>
+  ): Promise<T> {
+    return Promise.resolve(sendPromise)
+      .catch(reason => {
+        let errorMessage: string;
+        if (typeof reason === 'string') errorMessage = reason;
+        else if (
+          typeof reason === 'object' &&
+          Object.prototype.hasOwnProperty.call(reason, 'message')
+        )
+          errorMessage = reason.message;
+
+        spans.forEach(span =>
+          span.setStatus({
+            code: SpanStatusCode.ERROR,
+            message: errorMessage,
+          })
+        );
+
+        throw reason;
+      })
+      .finally(() => {
+        spans.forEach(span => span.end());
+      });
+  }
+
+  private _addModuleVersion(span: Span) {
+    if (this._config.moduleVersionAttributeName === undefined) {
+      return;
+    }
+
+    if (this.moduleVersion === undefined) {
+      return;
+    }
+
+    span.setAttribute(
+      this._config.moduleVersionAttributeName,
+      this.moduleVersion
+    );
+  }
+
+  private _startConsumerSpan(
+    topic: string,
+    message: KafkaMessage | undefined,
+    operation: string,
+    context: Context | undefined,
+    link?: Link
+  ) {
+    const span = this.tracer.startSpan(
+      topic,
+      {
+        kind: SpanKind.CONSUMER,
+        attributes: {
+          [SEMATTRS_MESSAGING_SYSTEM]: 'kafka',
```

Comment:
Could we use constants from SemConv here?

Reply:
I saw that `SemanticAttributes` has a deprecation notice, so I converted to these constants. Or did you have something else in mind?

Reply:
I'm confused. @Flarna The code has:

```js
import {
  MESSAGINGDESTINATIONKINDVALUES_TOPIC,
  MESSAGINGOPERATIONVALUES_PROCESS,
  MESSAGINGOPERATIONVALUES_RECEIVE,
  SEMATTRS_MESSAGING_SYSTEM,
  SEMATTRS_MESSAGING_DESTINATION_KIND,
  SEMATTRS_MESSAGING_DESTINATION,
  SEMATTRS_MESSAGING_OPERATION,
} from '@opentelemetry/semantic-conventions';
```

above, so constants from semconv *are* being used here, right?

Reply:
In [SemConv](https://github.com/open-telemetry/semantic-conventions/blob/v1.25.0/docs/messaging/messaging-spans.md) there are values defined for well known messaging systems like `kafka`. I was assuming there are generated constants for it.
So it's not about the keys it's about value `kafka`.

Reply:
Okay, understood.

Given that the JS `@opentelemetry/semantic-conventions` package is way back on v1.7.0, semconv did not define well-known messaging systems at that version. IIUC, a set of well-known messaging systems was only added about 5mo ago in https://github.com/open-telemetry/semantic-conventions/pull/517 -- part of semconv v1.24.0 (2023-12-15).

The Java semconv package has this class for that enum:
https://github.com/open-telemetry/semantic-conventions-java/blob/7da24068eea69dff11a78d59750b115dc4c5854d/semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/MessagingIncubatingAttributes.java#L369-L401
but AFAICT there is no Java code in https://github.com/open-telemetry/opentelemetry-java-instrumentation using them yet.

Reply:
`kafka` is not in the semconvs package yet, or am I missing something?

Reply:
@seemk I don't think you are missing anything, no. `kafka` isn't defined in the current `@opentelemetry/semantic-conventions` package as a string value.

Reply:
Correct, it's not included. I was assuming it is there because it is in spec.
Completely forgot out the fact that the version used here is several years old where `kafka` as not yet part of the spec.